### PR TITLE
Wrapper expire

### PR DESCRIPTION
**Pull request by jefflau**

### CODE REVIEW

File: contracts/wrapper/NameWrapper.sol [hunk 1]
```diff
@@ -514,9 +542,10 @@ contract NameWrapper is
      * @param node The namehash of the name to check fuses on.
      * @param fuseMask A bitmask of fuses that must not be burned.
      */
-    modifier operationAllowed(bytes32 node, uint96 fuseMask) {
-        (, uint96 fuses) = getData(uint256(node));
-        if (fuses & fuseMask != 0) {
+
+    modifier operationAllowed(bytes32 node, uint32 fuseMask) {
+        (, uint32 fuses, uint64 expiry) = getData(uint256(node));
+        if (fuses & fuseMask != 0 && expiry > block.timestamp) {
```
Comment (Arachnid) [inline contracts/wrapper/NameWrapper.sol:548]: Doesn't `getData` already set fuses to 0 if we're after the expiry?

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: contracts/wrapper/NameWrapper.sol [hunk 2]
```diff
@@ -210,19 +211,19 @@ contract NameWrapper is
      * @param wrappedOwner The owner of the wrapped name.
      * @param duration The duration, in seconds, to register the name for.
      * @param resolver The resolver address to set on the ENS registry (optional).
-     * @return expires The expiry date of the new name, in seconds since the Unix epoch.
+     * @return registrarExpiry The expiry date of the new name, in seconds since the Unix epoch.
```
[the remark this answers is not in the document]
Comment (jefflau) [reply contracts/wrapper/NameWrapper.sol:214]: I wanted to differentiate between expiry (for the wrapper) and expiry (for the base registrar). Calling it expiry makes it exactly the same as the wrapper expiry

[the remark this answers is not in the document]
Comment (jefflau) [reply contracts/wrapper/NameWrapper.sol:214]: I wanted to differentiate between expiry (for the wrapper) and expiry (for the base registrar). Calling it expiry makes it exactly the same as the wrapper expiry

File: contracts/wrapper/NameWrapper.sol [hunk 3]
```diff
@@ -263,6 +272,9 @@ contract NameWrapper is
             revert IncompatibleParent();
         }
 
+        // PARENT_CANNOT_CONTROL can only be burnt by the parent using setChildFuses() or setSubnodeOwner/Record()
+        _checkFusesForParentCannotControl(node, fuses);
```
[the remark this answers is not in the document]
Comment (jefflau) [reply contracts/wrapper/NameWrapper.sol:276]: Yeah this makes sense. I'll remove fuses from the wrap function

File: contracts/wrapper/NameWrapper.sol [hunk 4]
```diff
@@ -322,73 +334,86 @@ contract NameWrapper is
...
+        (address owner, uint32 oldFuses, uint64 expiry) = getData(
+            uint256(node)
+        );
+
+        _checkFusesForParentCannotControl(node, fuses);
 
-        _burnFuses(node, _fuses);
+        (, , uint64 maxExpiry) = getData(uint256(parentNode));
+        if (expiry > maxExpiry) {
+            // check if parent has burned CANNOT_UNWRAP when you're burning any fuses
```
[the remark this answers is not in the document]
Comment (jefflau) [reply contracts/wrapper/NameWrapper.sol:355]: Yes this makes sense. We don't need any of this. I'll refactor this function to get rid of the parentNode

File: contracts/wrapper/NameWrapper.sol (hunk 1 above)
Comment (jefflau) [reply contracts/wrapper/NameWrapper.sol:548]: Yes this is true. I'll remove `expiry > block.timestamp`

File: contracts/wrapper/NameWrapper.sol [hunk 5]
```diff
@@ -232,27 +233,35 @@ contract NameWrapper is
...
+        uint256 duration,
+        uint64 expiry
+    ) external override onlyController returns (uint256 expires) {
+        bytes32 node = _makeNode(ETH_NODE, bytes32(tokenId));
+        address owner;
+        uint32 fuses;
+        uint64 oldExpiry;
+
+        expires = registrar.renew(tokenId, duration);
+        (owner, fuses, oldExpiry) = getData(uint256(node));
```
Comment (Arachnid) [inline contracts/wrapper/NameWrapper.sol:247]:
```suggestion

        expires = registrar.renew(tokenId, duration);
        (address owner, uint32 fuses, uint64 oldExpiry) = getData(uint256(node));
```

**Review by Arachnid — APPROVED** [review]

File: contracts/wrapper/NameWrapper.sol [hunk 6]
```diff
@@ -322,73 +328,78 @@ contract NameWrapper is
...
+            }
+            // max expiry is set to the expiry on the registrar
+            maxExpiry = uint64(registrar.nameExpires(uint256(labelhash)));
+        } else {
+            if (!isTokenOwnerOrApproved(parentNode, msg.sender)) {
+                revert Unauthorised(node, msg.sender);
+            }
 
-        uint96 newFuses = fuses | _fuses;
+            // do not allow burning of fuses if PARENT_CANNOT_CONTROL hasn't been burnt
```
[the remark this answers is not in the document]
Comment (jefflau) [reply contracts/wrapper/NameWrapper.sol:371]: Yes I believe so! Will remove

File: contracts/wrapper/NameWrapper.sol (hunk 1 above)
Comment (BHAIsandeep) [reply contracts/wrapper/NameWrapper.sol]: I'm back loss my all wallet key please help

**Pull request merged**